magma/array.py

```diff
@@ -228,7 +241,7 @@ def __init__(self, *args, **kwargs):
         else:
             for i in range(self.N):
                 T = self.T
-                t = T(name=ArrayRef(self, i))
+                t = magma_type(T)(name=ArrayRef(self, i))
```

Comment:
is this the desired behavior? this excludes the possibility of `Array[N, MyProtocol]()[0]` returning an instance of `MyProtocol`? (Admittedly, it probably wasn't possible before either.)

Reply:
tried to get MagmaProtocol one and failed. let me try again.

Reply:
tried like this, 

    t = T._from_magma_value_(T._to_magma_()(name=ArrayRef(self, i)))

but no luck..

        def _get_magma_value_(self):
            # Return output
    >       return self.circuit()
    E       TypeError: __call__() missing 1 required positional argument: 'output'

Reply:
Can you provide the full stacktrace of this error? not sure exactly where this is coming from

Reply:
Ok, figured out what's going on.
- patch https://gist.github.com/splhack/e97dc61a886626f956fecfb9258b742a
- full failure log https://gist.github.com/splhack/27bdefdb19f95eca9d7b05da8805e867

I'm using sequential2 test, so all `MagmaProtocol` class is `_SequentialRegisterWrapper`. The class `__init__` expects always circuit (`Register`) as magma value. `_get_magma_value_` calls `__call__()` method of the "circuit".

But, in the failed tests, the magma value is non-circuit, like `Bit`. It doesn't have `__call__()`.

Reply:
It seems like `_SequentialRegisterWrapperMeta._to_magma_` needs to be updated then?

Reply:
These code paths 

- array.py https://github.com/phanrahan/magma/blob/69c64ce286cb89a6cd10285d4a0c1cb71faf96b5/magma/array.py#L229-L232
- tuple.py https://github.com/phanrahan/magma/blob/2b16f67f687ff7a59860e95960325bb8d85c7476/magma/tuple.py#L134-L138

are only called from interface.py `_make_port` (in my understanding.)

- interface.py https://github.com/phanrahan/magma/blob/a7e30f265137cb1e5e87bff6580a6c2ffd265ca3/magma/interface.py#L63

but

> this excludes the possibility of Array\[N, MyProtocol\]()[0] returning an instance of MyProtocol?

No, it can be done by the other side of the code path. For example, from this PR tests,

    def __init__(self):
        self.v = m.Register(T=m.Bit, init=m.bit(0))()
    def __call__(self):
        a = m.array([self.v.prev(), self.v.prev()])
        print(type(a)) # -> Array[2, _SequentialRegisterWrapperOut(Bit)]
        print(isinstance(a[0], MagmaProtocol)) # -> True